core/src/main/java/org/apache/calcite/sql/type/ReturnTypes.java

```diff
@@ -655,6 +655,54 @@ public int size() {
         return ret;
       };
 
+
+  /**
+   * Type-inference strategy for String concatenation.
```

Comment:
Improve the comment, you can refer to `DYADIC_STRING_SUM_PRECISION` for "||" operator.